File: run_demo_transcript.ts
import {
  Proposal,
  Evidence,
  NegotiationSimulator
} from "./src/reflection_memory";

export function runTranscript(): void {
  console.log("=" .repeat(80));
  console.log("  DECISION TWIN SYSTEM: REFLECTION & MEMORY NEGOTIATION TRANSCRIPT");
  console.log("  SCENARIO: Industry 4.0 Manufacturing & Machine #4 Optimization");
  console.log("=" .repeat(80));
  console.log("Target Branch: feat/reflection-memory");
  console.log("Agents Involved: Proposer, SafetyAgent, HistoricalMemoryAgent (RAG), DevilsAdvocateAgent");
  console.log("=" .repeat(80) + "\n");

  const simulator = new NegotiationSimulator();

  // ---------------------------------------------------------
  // ROUND 1: INITIAL PROPOSAL (Will trigger Safety Hard Veto & DA Challenge)
  // ---------------------------------------------------------
  console.log("┌" + "─".repeat(78) + "┐");
  console.log("│ ROUND 1: INITIAL PROPOSAL (Machine #4 High-Speed Overhaul & Throughput Boost)│");
  console.log("└" + "─".repeat(78) + "┘");

  const propR1: Proposal = {
    id: "PROP-001",
    title: "Machine #4 High-Speed Overhaul & Throughput Boost",
    description: "Upgrade Machine #4 CNC spindle drive and conveyor feed directly to 4.5x overdrive capacity to meet aggressive production targets.",
    proposed_by: "PlantOps_Lead",
    parameters: {
      budget_allocation: 120000.0,
      traffic_spike_multiplier: 4.5,
      auto_failover: false,
      prohibited_flags: []
    },
    evidences: [
      {
        id: "EV-PRELIM-01",
        description: "Supplier marketing spec flyer",
        source: "ToolingVendorCorp",
        confidence_score: 0.45
      }
    ],
    risk_level: "HIGH"
  };

  console.log(`[PROPOSAL SUBMITTED]: '${propR1.title}'`);
  console.log(`  Proposed By: ${propR1.proposed_by}`);
  console.log(`  Budget Requested: $${(propR1.parameters["budget_allocation"] as number).toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`);
  console.log(`  Production Speed Multiplier: ${propR1.parameters["traffic_spike_multiplier"]}x`);
  console.log(`  Auto-Interlock / Safety Failover: ${propR1.parameters["auto_failover"]}`);
  console.log(`  Evidence Count: ${propR1.evidences!.length} (Confidence: ${propR1.evidences![0].confidence_score})`);

  const resR1 = simulator.evaluateProposalRound(1, propR1);

  console.log("\n[SAFETY AGENT EVALUATION]:");
  console.log(`  Verdict: ${resR1.safety_verdict!.verdict}`);
  console.log(`  Details: ${resR1.safety_verdict!.details}`);
  for (const rule of resR1.safety_verdict!.violated_rules) {
    console.log(`    ❌ Violated Rule: ${rule}`);
  }

  console.log("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:");
  for (const inc of resR1.retrieved_incidents!) {
    console.log(`  🔍 Past Incident Match: [${inc.id}] ${inc.title} (Domain: ${inc.domain})`);
    console.log(`     Outcome: ${inc.outcome}`);
  }

  console.log("\n[DEVIL'S ADVOCATE EVALUATION]:");
  console.log(`  Challenge Summary: ${resR1.counter_argument!.challenge_summary}`);
  console.log(`  Calculated Risk Score: ${resR1.counter_argument!.risk_score.toFixed(2)}`);
  console.log("  Missing Evidence Demanded:");
  for (const me of resR1.counter_argument!.missing_evidences) {
    console.log(`    ⚠️ ${me}`);
  }
  console.log("  Counter-Claims Raised:");
  for (const cc of resR1.counter_argument!.counter_claims) {
    console.log(`    ⚡ ${cc}`);
  }

  console.log(`\n>>> ROUND 1 STATUS: ${resR1.status} <<<\n`);

  // ---------------------------------------------------------
  // ROUND 2: REVISED PROPOSAL 1 (Fixes Budget, but triggers DA & RAG Memory Challenge)
  // ---------------------------------------------------------
  console.log("┌" + "─".repeat(78) + "┐");
  console.log("│ ROUND 2: REVISED PROPOSAL (Scaled-Down Budget, Unverified Scaling)       │");
  console.log("└" + "─".repeat(78) + "┘");

  const propR2: Proposal = {
    id: "PROP-001-REV1",
    title: "Machine #4 Static Overhaul with Disabled Sensor Replicas",
    description: "Reduced tooling budget to $45,000 to meet SOP limit, keeping redundant PLC sensor feeds disabled to save setup cost.",
    proposed_by: "PlantOps_Lead",
    parameters: {
      budget_allocation: 45000.0,
      traffic_spike_multiplier: 3.5,
      auto_failover: true,
      prohibited_flags: ["disable_replicas"]
    },
    evidences: [
      {
        id: "EV-STAGING-01",
        description: "Internal shop floor test log",
        source: "PlantMetrics",
        confidence_score: 0.68
      }
    ],
    risk_level: "MEDIUM"
  };

  console.log(`[REVISED PROPOSAL SUBMITTED]: '${propR2.title}'`);
  console.log(`  Budget Requested: $${(propR2.parameters["budget_allocation"] as number).toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })} (Complies with $50k SOP limit)`);
  console.log(`  Production Speed Multiplier: ${propR2.parameters["traffic_spike_multiplier"]}x`);

  const resR2 = simulator.evaluateProposalRound(2, propR2);

  console.log("\n[SAFETY AGENT EVALUATION]:");
  console.log(`  Verdict: ${resR2.safety_verdict!.verdict}`);
  console.log(`  Details: ${resR2.safety_verdict!.details}`);
  for (const rule of resR2.safety_verdict!.violated_rules) {
    console.log(`    ❌ Violated Rule: ${rule}`);
  }

  console.log("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:");
  for (const inc of resR2.retrieved_incidents!) {
    console.log(`  🔍 Past Incident Match: [${inc.id}] ${inc.title}`);
    console.log(`     Lessons Learned: ${inc.lessons_learned[0] || "N/A"}`);
  }

  console.log("\n[DEVIL'S ADVOCATE EVALUATION]:");
  console.log(`  Challenge Summary: ${resR2.counter_argument!.challenge_summary}`);
  console.log(`  Calculated Risk Score: ${resR2.counter_argument!.risk_score.toFixed(2)}`);
  for (const me of resR2.counter_argument!.missing_evidences) {
    console.log(`    ⚠️ ${me}`);
  }
  for (const cc of resR2.counter_argument!.counter_claims) {
    console.log(`    ⚡ ${cc}`);
  }

  console.log(`\n>>> ROUND 2 STATUS: ${resR2.status} <<<\n`);

  // ---------------------------------------------------------
  // ROUND 3: FINAL REVISED PROPOSAL (Fully Compliant & Validated -> RESOLVED)
  // ---------------------------------------------------------
  console.log("┌" + "─".repeat(78) + "┐");
  console.log("│ ROUND 3: FINAL REVISED PROPOSAL (Predictive Maintenance & Staged Rollout)│");
  console.log("└" + "─".repeat(78) + "┘");

  const propR3: Proposal = {
    id: "PROP-001-REV2",
    title: "Machine #4 Predictive Maintenance & Staged Conveyor Rollout",
    description: "Deploy Machine #4 with 20% staged conveyor load, active redundant PLC sensors, dynamic speed scaling up to 2.5x, and automated thermal safety interlocks.",
    proposed_by: "PlantOps_Lead",
    parameters: {
      budget_allocation: 38000.0,
      traffic_spike_multiplier: 2.5,
      auto_failover: true,
      prohibited_flags: []
    },
    evidences: [
      {
        id: "EV-LOAD-99",
        description: "Machine #4 3x Peak Load Stress Benchmark",
        source: "AutomatedTestRig",
        confidence_score: 0.96
      }
    ],
    risk_level: "LOW"
  };

  console.log(`[FINAL PROPOSAL SUBMITTED]: '${propR3.title}'`);
  console.log(`  Budget Requested: $${(propR3.parameters["budget_allocation"] as number).toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`);
  console.log(`  Production Speed Multiplier: ${propR3.parameters["traffic_spike_multiplier"]}x`);
  console.log(`  Evidence: ${propR3.evidences![0].id} - ${propR3.evidences![0].description} (Confidence: ${propR3.evidences![0].confidence_score})`);

  const resR3 = simulator.evaluateProposalRound(3, propR3);

  console.log("\n[SAFETY AGENT EVALUATION]:");
  console.log(`  Verdict: ${resR3.safety_verdict!.verdict}`);
  console.log(`  Details: ${resR3.safety_verdict!.details}`);

  console.log("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:");
  for (const inc of resR3.retrieved_incidents!) {
    console.log(`  🔍 Past Precedent Found: [${inc.id}] ${inc.title}`);
    console.log(`     Outcome: ${inc.outcome}`);
  }

  console.log("\n[DEVIL'S ADVOCATE EVALUATION]:");
  console.log(`  Challenge Summary: ${resR3.counter_argument!.challenge_summary}`);
  console.log(`  Calculated Risk Score: ${resR3.counter_argument!.risk_score.toFixed(2)}`);

  console.log(`\n>>> ROUND 3 STATUS: ${resR3.status} <<<`);
  console.log("=" .repeat(80));
  console.log("       NEGOTIATION CONVERGED & RESOLVED SUCCESSFULLY!");
  console.log("=" .repeat(80) + "\n");
}

if (require.main === module) {
  runTranscript();
}

File: run_demo_transcript.py
import sys
import os
import json

# Ensure stdout handles UTF-8 on Windows
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

# Ensure src is in python path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "src")))

from reflection_memory import (
    Proposal,
    Evidence,
    DevilsAdvocateAgent,
    HistoricalMemoryAgent,
    SafetyAgent,
    NegotiationSimulator,
    VerdictType
)

def run_transcript():
    print("=" * 80)
    print("      DECISION TWIN SYSTEM: REFLECTION & MEMORY NEGOTIATION TRANSCRIPT")
    print("=" * 80)
    print("Target Branch: feat/reflection-memory")
    print("Agents Involved: Proposer, SafetyAgent, HistoricalMemoryAgent (RAG), DevilsAdvocateAgent")
    print("=" * 80 + "\n")

    simulator = NegotiationSimulator()

    # ---------------------------------------------------------
    # ROUND 1: INITIAL PROPOSAL (Will trigger Safety Hard Veto & DA Challenge)
    # ---------------------------------------------------------
    print("┌" + "─" * 78 + "┐")
    print("│ ROUND 1: INITIAL PROPOSAL (Aggressive Cloud Infrastructure Redesign)      │")
    print("└" + "─" * 78 + "┘")

    prop_r1 = Proposal(
        id="PROP-001",
        title="Instant Global Cloud Migration & Capacity Expansion",
        description="Migrate core billing service directly to unpartitioned high-mem instances to handle anticipated 4x flash sale spike.",
        proposed_by="LeadArch_Dev",
        parameters={
            "budget_allocation": 120000.0,
            "traffic_spike_multiplier": 4.5,
            "auto_failover": False,
            "prohibited_flags": []
        },
        evidences=[
            Evidence("EV-PRELIM-01", "Vendor marketing benchmark flyer", "CloudProviderCorp", 0.45)
        ],
        risk_level="HIGH"
    )

    print(f"[PROPOSAL SUBMITTED]: '{prop_r1.title}'")
    print(f"  Proposed By: {prop_r1.proposed_by}")
    print(f"  Budget Requested: ${prop_r1.parameters['budget_allocation']:,.2f}")
    print(f"  Traffic Multiplier: {prop_r1.parameters['traffic_spike_multiplier']}x")
    print(f"  Auto-Failover: {prop_r1.parameters['auto_failover']}")
    print(f"  Evidence Count: {len(prop_r1.evidences)} (Confidence: {prop_r1.evidences[0].confidence_score})")

    res_r1 = simulator.evaluate_proposal_round(1, prop_r1)

    print("\n[SAFETY AGENT EVALUATION]:")
    print(f"  Verdict: {res_r1.safety_verdict.verdict.value}")
    print(f"  Details: {res_r1.safety_verdict.details}")
    for rule in res_r1.safety_verdict.violated_rules:
        print(f"    ❌ Violated Rule: {rule}")

    print("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:")
    for inc in res_r1.retrieved_incidents:
        print(f"  🔍 Past Incident Match: [{inc.id}] {inc.title} (Domain: {inc.domain})")
        print(f"     Outcome: {inc.outcome}")

    print("\n[DEVIL'S ADVOCATE EVALUATION]:")
    print(f"  Challenge Summary: {res_r1.counter_argument.challenge_summary}")
    print(f"  Calculated Risk Score: {res_r1.counter_argument.risk_score:.2f}")
    print("  Missing Evidence Demanded:")
    for me in res_r1.counter_argument.missing_evidences:
        print(f"    ⚠️ {me}")
    print("  Counter-Claims Raised:")
    for cc in res_r1.counter_argument.counter_claims:
        print(f"    ⚡ {cc}")

    print(f"\n>>> ROUND 1 STATUS: {res_r1.status} <<<\n")

    # ---------------------------------------------------------
    # ROUND 2: REVISED PROPOSAL 1 (Fixes Budget, but triggers DA & RAG Memory Challenge)
    # ---------------------------------------------------------
    print("┌" + "─" * 78 + "┐")
    print("│ ROUND 2: REVISED PROPOSAL (Scaled-Down Budget, Unverified Scaling)       │")
    print("└" + "─" * 78 + "┘")

    prop_r2 = Proposal(
        id="PROP-001-REV1",
        title="Targeted Service Migration with Static Cluster",
        description="Reduced cluster size to fit budget cap, keeping read-replicas disabled to stay within cost.",
        proposed_by="LeadArch_Dev",
        parameters={
            "budget_allocation": 45000.0,
            "traffic_spike_multiplier": 3.5,
            "auto_failover": True,
            "prohibited_flags": ["disable_replicas"]
        },
        evidences=[
            Evidence("EV-STAGING-01", "Internal staging cluster metric logs", "DevOpsMetrics", 0.68)
        ],
        risk_level="MEDIUM"
    )

    print(f"[REVISED PROPOSAL SUBMITTED]: '{prop_r2.title}'")
    print(f"  Budget Requested: ${prop_r2.parameters['budget_allocation']:,.2f} (Complies with $50k SOP limit)")
    print(f"  Traffic Multiplier: {prop_r2.parameters['traffic_spike_multiplier']}x")

    res_r2 = simulator.evaluate_proposal_round(2, prop_r2)

    print("\n[SAFETY AGENT EVALUATION]:")
    print(f"  Verdict: {res_r2.safety_verdict.verdict.value}")
    print(f"  Details: {res_r2.safety_verdict.details}")
    for rule in res_r2.safety_verdict.violated_rules:
        print(f"    ❌ Violated Rule: {rule}")

    print("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:")
    for inc in res_r2.retrieved_incidents:
        print(f"  🔍 Past Incident Match: [{inc.id}] {inc.title}")
        print(f"     Lessons Learned: {inc.lessons_learned[0] if inc.lessons_learned else 'N/A'}")

    print("\n[DEVIL'S ADVOCATE EVALUATION]:")
    print(f"  Challenge Summary: {res_r2.counter_argument.challenge_summary}")
    print(f"  Calculated Risk Score: {res_r2.counter_argument.risk_score:.2f}")
    for me in res_r2.counter_argument.missing_evidences:
        print(f"    ⚠️ {me}")
    for cc in res_r2.counter_argument.counter_claims:
        print(f"    ⚡ {cc}")

    print(f"\n>>> ROUND 2 STATUS: {res_r2.status} <<<\n")

    # ---------------------------------------------------------
    # ROUND 3: FINAL REVISED PROPOSAL (Fully Compliant & Validated -> RESOLVED)
    # ---------------------------------------------------------
    print("┌" + "─" * 78 + "┐")
    print("│ ROUND 3: FINAL REVISED PROPOSAL (Canary Auto-Scale & Load Tested)        │")
    print("└" + "─" * 78 + "┘")

    prop_r3 = Proposal(
        id="PROP-001-REV2",
        title="Canary Blue-Green Migration with Dynamic Auto-Scaling",
        description="Deploy 20% canary traffic with active read-replicas, dynamic auto-scaling up to 3x, and automated instant rollbacks.",
        proposed_by="LeadArch_Dev",
        parameters={
            "budget_allocation": 38000.0,
            "traffic_spike_multiplier": 2.5,
            "auto_failover": True,
            "prohibited_flags": []
        },
        evidences=[
            Evidence("EV-LOAD-99", "Synthetic 3x Peak Load Test Benchmark", "AutomatedPerfSuite", 0.96)
        ],
        risk_level="LOW"
    )

    print(f"[FINAL PROPOSAL SUBMITTED]: '{prop_r3.title}'")
    print(f"  Budget Requested: ${prop_r3.parameters['budget_allocation']:,.2f}")
    print(f"  Traffic Multiplier: {prop_r3.parameters['traffic_spike_multiplier']}x")
    print(f"  Evidence: {prop_r3.evidences[0].id} - {prop_r3.evidences[0].description} (Confidence: {prop_r3.evidences[0].confidence_score})")

    res_r3 = simulator.evaluate_proposal_round(3, prop_r3)

    print("\n[SAFETY AGENT EVALUATION]:")
    print(f"  Verdict: {res_r3.safety_verdict.verdict.value}")
    print(f"  Details: {res_r3.safety_verdict.details}")

    print("\n[MID-REASONING HISTORICAL MEMORY LOOKUP]:")
    for inc in res_r3.retrieved_incidents:
        print(f"  🔍 Past Precedent Found: [{inc.id}] {inc.title}")
        print(f"     Outcome: {inc.outcome}")

    print("\n[DEVIL'S ADVOCATE EVALUATION]:")
    print(f"  Challenge Summary: {res_r3.counter_argument.challenge_summary}")
    print(f"  Calculated Risk Score: {res_r3.counter_argument.risk_score:.2f}")

    print(f"\n>>> ROUND 3 STATUS: {res_r3.status} <<<")
    print("=" * 80)
    print("       NEGOTIATION CONVERGED & RESOLVED SUCCESSFULLY!")
    print("=" * 80 + "\n")

if __name__ == "__main__":
    run_transcript()
